synapse/util/async_helpers.py

```diff
@@ -541,24 +543,40 @@ def write(self, key: str) -> AsyncContextManager:
         if curr_writer:
             to_wait_on.append(curr_writer)
 
-        # We can clear the list of current readers since the new writer waits
+        # We can clear the list of current readers since `new_defer` waits
         # for them to finish.
         curr_readers.clear()
         self.key_to_current_writer[key] = new_defer
 
         @asynccontextmanager
         async def _ctx_manager() -> AsyncIterator[None]:
+            to_wait_on_defer = defer.gatherResults(to_wait_on)
             try:
-                await make_deferred_yieldable(defer.gatherResults(to_wait_on))
+                # Wait for all current readers and the latest writer to finish.
+                # May raise a `CancelledError` if the `Deferred` wrapping us is
+                # cancelled.
+                await make_deferred_yieldable(to_wait_on_defer)
                 yield
             finally:
-                with PreserveLoggingContext():
-                    new_defer.callback(None)
-                # `self.key_to_current_writer[key]` may be missing if there was another
-                # writer waiting for us and it completed entirely within the
-                # `new_defer.callback()` call above.
-                if self.key_to_current_writer.get(key) == new_defer:
-                    self.key_to_current_writer.pop(key)
+
+                def release() -> None:
+                    with PreserveLoggingContext():
```

Comment:
I'm not wholly convinced the reasoning applies to `defer.callback()` since it's fairly "low level".

We do wrap `.callback(...)` calls in `PreserveLoggingContext` [elsewhere in Synapse](https://github.com/matrix-org/synapse/blob/release-v1.53/synapse/util/batching_queue.py#L164-L166).

More concretely, I believe that if there is a coroutine waiting on `defer`, then inside `defer.callback(...)`:
1. The coroutine's logging context will be restored because the Deferred went through `make_deferred_yieldable`. The logging context from outside the `defer.callback()` call will be forgotten.
2. The coroutine will run for a bit.
3. Then either:
    a) await another `make_deferred_yieldable` Deferred and restore the sentinel logging context, or
    b) terminate, at which point it will have the logging context it had when it was originally started.

Reply:
hrm, yeah, I think you're right. I was always of the belief that wrapping `.callback` with `PreserveLoggingContext` was cargo-culting, but your reasoning seems good.